scripts/publish.ts

```diff
@@ -0,0 +1,104 @@
+    console.error("The first argument must be either `stable` or `unstable`.");
+    process.exit(1);
+  }
+  const specifiers = args._.slice(1);
+  if (mode === "stable" && specifiers.length === 0) {
+    console.error(
+      "You must specify at least one package version (using the `<name>@<version>` format) to create a stable release.",
+    );
+    process.exit(1);
+  } else if (mode === "unstable" && specifiers.length > 0) {
+    console.error(
+      "You can't specify any package version when creating an unstable release.",
+    );
+    process.exit(1);
+  }
+
+  if (mode === "stable") {
+    const workspaceNames = await getWorkspaceNames();
+    const packageNameToSpecifiedVersion = new Map(
+      specifiers.map((specifier) => {
+        const parsedSpecifier = parsePackageNameSpecifier(specifier);
+        if (parsedSpecifier === undefined) {
+          console.error(
+            `Invalid package specifier: \`${specifier}\`. You must use the \`<name>@<version>\` format.`,
+          );
+          process.exit(1);
+        }
+        const [packageName, packageVersion] = parsedSpecifier;
+        if (!workspaceNames.has(packageName)) {
+          console.error(
+            `Invalid package specifier: \`${specifier}\`. \`${packageName}\` must be a non-private workspace.`,
+          );
+          process.exit(1);
+        }
+        const validPackageVersion = semverValid(packageVersion);
+        if (validPackageVersion === null) {
+          console.error(
+            `Invalid package specifier: \`${specifier}\`. Version \`${packageVersion}\` must be a valid SemVer string.`,
+          );
+          process.exit(1);
+        } else if (semverPrerelease(validPackageVersion) !== null) {
+          console.error(
+            `Invalid package specifier: \`${specifier}\`. Version \`${validPackageVersion}\` mustn't include any prerelease component.`,
+          );
+          process.exit(1);
+        }
+        return [packageName, validPackageVersion];
+      }),
+    );
+    console.log(packageNameToSpecifiedVersion); // TODO: remove this line.
+  }
+
+  const { stdout: status } = await $`git status --porcelain`;
+  if (status.length > 0) {
+    console.error("This command must be executed on a clean repository.");
+    process.exit(1);
+  } else if (mode === "stable") {
+    const { stdout: currentBranchName } = await $`git branch --show-current`;
+    if (currentBranchName !== defaultBranchName) {
+      console.error(
+        `You must be on the \`${defaultBranchName}\` branch to create a stable release.`,
+      );
+      process.exit(1);
+    }
+  }
+}
+
+async function getWorkspaceNames(): Promise<Set<string>> {
+  const { stdout } = await $`yarn workspaces list --json --no-private`;
+  return new Set(
+    stdout
+      .split("\n")
+      .map((line) => (JSON.parse(line) as { name: string }).name),
+  );
+}
+
+function parsePackageNameSpecifier(
+  specifier: string,
+): [name: string, version: string] | undefined {
```

Comment:
Why do you choose an array return value rather than an object? I don't see any problem with it, I'm just curious if there's a benefit we are missing out on elsewhere.